### ISSUE TITLE
How to run the ENAS Tutorial on a GPU?

### ISSUE THREAD

**Issue by andyDoucette** [awslabs/autogluon#563]
## Problem
I'm working on trying to get the [ENAS Tutorial](https://autogluon.mxnet.io/tutorials/nas/enas_proxylessnas.html) to run on a GPU, but every time I do this, I get an obscure CUDA initialization error.  The error message is quite opaque, and I'm not sure where the problem lies.  I would really appreciate some help if you're willing.

Here is the code:

```python
#Docs https://autogluon.mxnet.io/tutorials/nas/enas_proxylessnas.html
import autogluon as ag

from autogluon.task.image_classification.dataset import get_built_in_dataset

import mxnet as mx

#Docs http://34.201.8.176/versions/1.4.1/api/python/gluon/nn.html
import mxnet.gluon.nn as nn
from autogluon.contrib.enas import *


#An overview of gluon:
#http://34.201.8.176/versions/1.4.1/tutorials/gluon/gluon.html?highlight=block

class Identity(mx.gluon.HybridBlock):
    def hybrid_forward(self, F, x):
        return x

class ConvBNReLU(mx.gluon.HybridBlock):
    def __init__(self, in_channels, channels, kernel, stride):
        super().__init__()
        
        #All layers must be attributes, since this derrives from a mx.gluon.Block:
        padding = (kernel - 1) // 2
        self.conv = nn.Conv2D(channels, kernel, stride, padding, in_channels=in_channels)
        self.bn = nn.BatchNorm(in_channels=channels)
        self.relu = nn.Activation('relu')
        
    def hybrid_forward(self, F, x):
        #Compose the layers to get our result layer:
        return self.relu(self.bn(self.conv(x)))
    
@enas_unit()
class ResUnit(mx.gluon.HybridBlock):
    def __init__(self, in_channels, channels, hidden_channels, kernel, stride):
        super().__init__()
        self.conv1 = ConvBNReLU(in_channels, hidden_channels, kernel, stride)
        self.conv2 = ConvBNReLU(hidden_channels, channels, kernel, 1)
        
        #Define our shortcut layer type:
        if in_channels == channels and stride == 1:
            self.shortcut = Identity()
        else:
            self.shortcut = nn.Conv2D(channels, 1, stride, in_channels=in_channels)
            
            
    def hybrid_forward(self, F, x):
        #Compose the layers to get our result layer.  Add the results, since only one should be
        #active at a time (I think):
        return self.conv2(self.conv1(x)) + self.shortcut(x)
    
def enas_example():
    '''
    The example from https://autogluon.mxnet.io/tutorials/nas/enas_proxylessnas.html
    '''
    
    #This is where the computations will be performed (cpu, gpu, etc):
    context = mx.gpu() if mx.context.num_gpus() else mx.cpu()
    
    mynet = ENAS_Sequential(
        ResUnit(1, 8, hidden_channels=ag.space.Categorical(4, 8), kernel=ag.space.Categorical(3, 5), stride=2),
        ResUnit(8, 8, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=2),
        ResUnit(8, 16, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=2),
        ResUnit(16, 16, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=1, with_zero=True),
        ResUnit(16, 16, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=1, with_zero=True),
        ResUnit(16, 16, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=1, with_zero=True),
        ResUnit(16, 16, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=1, with_zero=True),
        ResUnit(16, 16, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=1, with_zero=True),
        ResUnit(16, 16, hidden_channels=8, kernel=ag.space.Categorical(3, 5), stride=1, with_zero=True),
        nn.GlobalAvgPool2D(),
        nn.Flatten(),
        nn.Activation('relu'),
        nn.Dense(10, in_units=16),
    )

    mynet.initialize(ctx=context)

    #mynet.graph
    
    x = mx.nd.random.uniform(shape=(1, 1, 28, 28), ctx=context)
    y = mynet.evaluate_latency(x)
    
    print('Average latency is {:.2f} ms, latency of the current architecture is {:.2f} ms'.format(mynet.avg_latency, mynet.latency))
    
    print('Number of parameters is %i' %mynet.nparams)

    #Define the reward function:
    reward_fn = lambda metric, net: metric * ((net.avg_latency / net.latency) ** 0.1)

    #Where checkpoints are saved.  We need to set this so it doesn't clutter w/ a ./enas directory
    checkname='/app/data/.enas/checkpoint.ag'
    
    batch_size=128
    num_cpus=4
    
    #Get our datasets:
    dataset_name='mnist'
    
    #Get our datasets, and put them in the correct context:
    train_set = get_built_in_dataset(dataset_name, train=True, batch_size=batch_size,
                                             num_workers=num_cpus, shuffle=True)\
                    .transform(lambda x,y: to_context(context, x, y), lazy=False)
                
    val_set = get_built_in_dataset(dataset_name, train=False, batch_size=batch_size,
                                    num_workers=num_cpus, shuffle=True)\
                    .transform(lambda x,y: to_context(context, x, y), lazy=False)
    
    #TODO: Remove:
    test=train_set[0]
    
    #Construct experiment scheduler, which automatically creates an RL controller based on user-defined search space.
    scheduler = ENAS_Scheduler( mynet, train_set=train_set, val_set=val_set,
                                reward_fn=reward_fn, batch_size=batch_size, num_gpus=1,
                                warmup_epochs=0, epochs=1, controller_lr=3e-3,
                                plot_frequency=10, update_arch_frequency=5, checkname=checkname)
    scheduler.run()
    
    mynet.graph
    
def to_context(context, x, y):
    '''
    Puts a datset sample=(x,y) into a context (gpu, cpu, etc).
    '''

    x=x.copyto(context)
    
    return (x,y)
    
if __name__ == '__main__':
    print('running')
    enas_example()

```
As you can see, I'm being very careful to put both the model and the data in the gpu context.

Here is what I get when I run it on a system with a Nvidia Tesla M60 GPU, in a docker container based off of nvidia/cuda:10.1-runtime-ubuntu18.04
```
[user@ba boundryDetector]$ python3 ./enas_example.py 
/usr/local/lib/python3.6/dist-packages/mxnet/optimizer/optimizer.py:167: UserWarning: WARNING: New optimizer gluonnlp.optimizer.lamb.LAMB is overriding existing optimizer mxnet.optimizer.optimizer.LAMB
  Optimizer.opt_registry[name].__name__))
running
[02:41:36] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (set the environment variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
Average latency is 5.71 ms, latency of the current architecture is 7.20 ms
Number of parameters is 18410
get_built_in_dataset mnist
get_built_in_dataset mnist
  0%|                                                                                             | 0/1 [00:00<?, ?it/sterminate called after throwing an instance of 'dmlc::Error'                                     | 0/468 [00:00<?, ?it/s]
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]


terminate called after throwing an instance of 'dmlc::Error'
  what():  [02:41:58] /home/ubuntu/mxnet-distro/mxnet-build/3rdparty/mshadow/mshadow/./tensor_gpu-inl.h:35: Check failed: e == cudaSuccess: CUDA: initialization error
Stack trace:
  [bt] (0) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x6b8b5b) [0x7f67227f9b5b]
  [bt] (1) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37ab842) [0x7f67258ec842]
  [bt] (2) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c7dd5) [0x7f6725908dd5]
  [bt] (3) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c8136) [0x7f6725909136]
  [bt] (4) /usr/local/lib/python3.6/dist-packages/mxnet/libmxnet.so(+0x37c3114) [0x7f6725904114]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xbda50) [0x7f6703b66a50]
  [bt] (6) /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f678dd466db]
  [bt] (7) /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f678e07fa3f]
```

I'm using:
```
autogluon==0.0.11
mxnet==1.6.0
```

The same code works fine on a cpu-only system (because it doesn't even try to use cuda).

I'm not sure where to start debugging this since it's so low in the system.  Should this error be caught at a higher level so we can provide a more informative error message?  Do you know if the issue is more likely in my system setup (library compatibility, etc) or am I doing something wrong in using mxnet and autogluon?

I greatly appreciate any assistance you may offer.  :)

## Update
The [mxnet multi-gpu example](https://mxnet.apache.org/versions/1.5.0/tutorials/gluon/multi_gpu.html) works for me in the same environment, so that suggests to me that the problem isn't in my environment but is somewhere in the autogluon code or my use of it. 

```python
#From https://mxnet.apache.org/versions/1.5.0/tutorials/gluon/multi_gpu.html

import mxnet as mx

n_gpu = mx.context.num_gpus()
context = [mx.gpu(0), mx.gpu(1)] if n_gpu >= 2 else \
          [mx.gpu(), mx.gpu()] if n_gpu == 1 else \
          [mx.cpu(), mx.cpu()]

a = mx.nd.array([1, 2, 3], ctx=context[0])
b = mx.nd.array([5, 6, 7], ctx=context[1])

c = a + b.as_in_context(a.context)

from mxnet import init
from mxnet.gluon import nn

net = nn.Sequential()
net.add(nn.Conv2D(channels=6, kernel_size=5, activation='relu'),
        nn.MaxPool2D(pool_size=2, strides=2),
        nn.Conv2D(channels=16, kernel_size=3, activation='relu'),
        nn.MaxPool2D(pool_size=2, strides=2),
        nn.Flatten(),
        nn.Dense(120, activation="relu"),
        nn.Dense(84, activation="relu"),
        nn.Dense(10))

net.initialize(init=init.Xavier(), ctx=context)

data = mx.random.uniform(shape=(100, 10))
result = mx.gluon.utils.split_and_load(data, ctx_list=context)

train_data = mx.gluon.data.vision.MNIST(train=True).transform_first(mx.gluon.data.vision.transforms.ToTensor())
val_data = mx.gluon.data.vision.MNIST(train=False).transform_first(mx.gluon.data.vision.transforms.ToTensor())

batch_size = 128
train_loader = mx.gluon.data.DataLoader(train_data, shuffle=True, batch_size=batch_size)
val_loader = mx.gluon.data.DataLoader(val_data, shuffle=False, batch_size=batch_size)

trainer = mx.gluon.Trainer(
    params=net.collect_params(),
    optimizer='sgd',
    optimizer_params={'learning_rate': 0.04},
)

metric = mx.metric.Accuracy()
loss_function = mx.gluon.loss.SoftmaxCrossEntropyLoss()

num_epochs = 10

for epoch in range(num_epochs):
    for inputs, labels in train_loader:
        actual_batch_size = inputs.shape[0]
        # Split data among GPUs. Since split_and_load is a deterministic function
        # inputs and labels are going to be split in the same way between GPUs.
        inputs = mx.gluon.utils.split_and_load(inputs, ctx_list=context, even_split=False)
        labels = mx.gluon.utils.split_and_load(labels, ctx_list=context, even_split=False)

        # The forward pass and the loss computation need to be wrapped
        # in a `record()` scope to make sure the computational graph is
        # recorded in order to automatically compute the gradients
        # during the backward pass.
        with mx.autograd.record():
            outputs = [net(input_slice) for input_slice in inputs]
            losses = [loss_function(o, l) for o, l in zip(outputs, labels)]

        # Iterate over losses to compute gradients for each input slice
        for loss in losses:
            loss.backward()

        # update metric for each output
        for l, o in zip(labels, outputs):
            metric.update(l, o)

        # Update the parameters by stepping the trainer; the batch size
        # is required to normalize the gradients by `1 / batch_size`.
        trainer.step(batch_size=actual_batch_size, ignore_stale_grad=True)

    # Print the evaluation metric and reset it for the next epoch
    name, acc = metric.get()
    print('After epoch {}: {} = {}'.format(epoch + 1, name, acc))
    metric.reset()
```

Result of running:
```
[user@ba boundryDetector]$ python3 ./multi_gpu_example.py                                                                                                                          
Downloading /root/.mxnet/datasets/mnist/train-images-idx3-ubyte.gz from https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/mnist/train-images-idx3-ubyte.gz...                                                                                                                                                                                  
Downloading /root/.mxnet/datasets/mnist/train-labels-idx1-ubyte.gz from https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/mnist/train-labels-idx1-ubyte.gz...                                                                                                                                                                                  
Downloading /root/.mxnet/datasets/mnist/t10k-images-idx3-ubyte.gz from https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/mnist/t10k-images-idx3-ubyte.gz... 
Downloading /root/.mxnet/datasets/mnist/t10k-labels-idx1-ubyte.gz from https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/mnist/t10k-labels-idx1-ubyte.gz... 
[03:14:35] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (set the environment variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)                                                                                                                                   
[03:14:35] src/kvstore/././comm.h:744: only 0 out of 2 GPU pairs are enabled direct access. It may affect the performance. You can set MXNET_ENABLE_GPU_P2P=0 to turn it off       
[03:14:35] src/kvstore/././comm.h:753: ..
[03:14:35] src/kvstore/././comm.h:753: ..
After epoch 1: accuracy = 0.73135
After epoch 2: accuracy = 0.9279333333333334
After epoch 3: accuracy = 0.95245
After epoch 4: accuracy = 0.9621
After epoch 5: accuracy = 0.9681833333333333
After epoch 6: accuracy = 0.9711333333333333
After epoch 7: accuracy = 0.97455
After epoch 8: accuracy = 0.9767166666666667
After epoch 9: accuracy = 0.97845
After epoch 10: accuracy = 0.9798333333333333
```

**Comment by Innixma** [comment]
@zhreshold Could you take a look at this?

**Comment by andyDoucette** [comment]
Let me know if you have trouble reproducing.  I can provide a full Dockerfile if needed.

**Comment by uzhao** [comment]
I got same problem when I try

TabularPrediction.fit(
        train_data=xy,
        label="y",
        problem_type="multiclass",
        ngpus_per_trial=1,
        auto_stack=True, 
        num_bagging_sets=5
)

There is no problem when not using GPU or not using auto_stack

autogluon.__version__ = '0.0.11'
mxnet.__version__ = '1.6.0'

**Comment by Innixma** [comment]
> I got same problem when I try
> 
> TabularPrediction.fit(
> train_data=xy,
> label="y",
> problem_type="multiclass",
> ngpus_per_trial=1,
> auto_stack=True,
> num_bagging_sets=5
> )
> 
> There is no problem when not using GPU or not using auto_stack
> 
> autogluon.**version** = '0.0.11'
> mxnet.**version** = '1.6.0'

I believe this may be a separate issue, could you make a new issue and include the run logs? Additionally, please use autogluon==0.0.12 to ensure you are up to date. Do note that Tabular does not use GPU beyond the neural network model at present.

**Comment by uzhao** [comment]
> > I got same problem when I try
> > TabularPrediction.fit(
> > train_data=xy,
> > label="y",
> > problem_type="multiclass",
> > ngpus_per_trial=1,
> > auto_stack=True,
> > num_bagging_sets=5
> > )
> > There is no problem when not using GPU or not using auto_stack
> > autogluon.**version** = '0.0.11'
> > mxnet.**version** = '1.6.0'
> 
> I believe this may be a separate issue, could you make a new issue and include the run logs? Additionally, please use autogluon==0.0.12 to ensure you are up to date. Do note that Tabular does not use GPU beyond the neural network model at present.

I can create a new issue as update to 0.0.12 doesn't fix it. Also I think catboost also use GPU? In fact, it will be great if GPU is supported in all models available like lightgbm
https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/catboost/catboost_model.py#L150

**Comment by Innixma** [comment]
> > > I got same problem when I try
> > > TabularPrediction.fit(
> > > train_data=xy,
> > > label="y",
> > > problem_type="multiclass",
> > > ngpus_per_trial=1,
> > > auto_stack=True,
> > > num_bagging_sets=5
> > > )
> > > There is no problem when not using GPU or not using auto_stack
> > > autogluon.**version** = '0.0.11'
> > > mxnet.**version** = '1.6.0'
> > 
> > 
> > I believe this may be a separate issue, could you make a new issue and include the run logs? Additionally, please use autogluon==0.0.12 to ensure you are up to date. Do note that Tabular does not use GPU beyond the neural network model at present.
> 
> I can create a new issue as update to 0.0.12 doesn't fix it. Also I think catboost also use GPU? In fact, it will be great if GPU is supported in all models available like lightgbm
> https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/ml/models/catboost/catboost_model.py#L150

Sounds good regarding creating a new issue. Also, GPU support is being tracked in #263, and should be fairly easy to implement.

**Comment by andyDoucette** [comment]
Somehow the discussion got side-tracked to tabular stuff.  The original issue was with ENAS.  Did anyone try to reproduce this besides me?

**Comment by Innixma** [comment]
Apologies for the side-tracking :) , @zhreshold is on vacation this week.

@zhreshold : when you get back, would you be able to look into this?

**Comment by andyDoucette** [comment]
Thanks Nick!

**Comment by andyDoucette** [comment]
@zhreshold :  I hope you had a good vacation.  Can you take a look at this in the next day or two please?  I really need GPU acceleration for my project to be success.  Thanks so much. :)

**Comment by zhreshold** [comment]
@andyDoucette No worries, trying to look into the issue there. If you can provide a working dockerfile and launcher script I probably can help you out faster

**Comment by andyDoucette** [comment]
I'm so glad you're on it. :)  Yeah, I'd be glad to help you reproduce it.

Ok, on an g2.2xlarge aws instance with 50GB of storage and the Ubuntu 18.04 AMI installed, extract [this directory](https://drive.google.com/file/d/1rxtzuyR9RKzqGtCGFJR_XjFXkj_ZTb0G/view?usp=sharing), and run `cd enasTest; ./setup.sh` to set up the host.  `sudo reboot` to reboot the system.  Log back into the server.

Then, run `ba` to enter the docker container (after building).

From within the docker container, you can run `mutli_gpu_example.py` to get the working example that does not use ENAS but does use the GPU.  You can then run `enas_example.py` to get the error message trying to use ENAS with the GPU.

When I go through this procedure now, I get a different message that comes from importing mxnet.  

`Illegal instruction (core dumped)`

I had to use a different mxnet in this setup then I did before because mxnet-cu101==1.6.0 was no longer available.  I'll message again tomorrow when I figure why I'm an additional error now and I wasn't before.  If you're feeling brave, feel free to see if you can start where I left off and get us the rest of the way to reproducing the issue.

**Comment by zhreshold** [comment]
@andyDoucette  You can still find the wheels for cu101 of mxnet 1.6.0 here: https://dist.mxnet.io/python/cu101

In particular: https://repo.mxnet.io/dist/python/cu101/mxnet_cu101-1.6.0-py2.py3-none-manylinux1_x86_64.whl

**Comment by zhreshold** [comment]
@andyDoucette 
Okay, I think I found a way to fix the issue, generally speaking, you can't intialize the network on gpu before forking the processes.
That's why there's a internal reset_context for you if `num_gpu > 0`: https://github.com/awslabs/autogluon/blob/master/autogluon/contrib/enas/enas_scheduler.py#L79

To fix your original example code, simply change:
```python
context = mx.gpu() if mx.context.num_gpus() else mx.cpu()
``` 
to 
```python
context = mx.cpu()
```

and you have already set the num_gpu in
```python
scheduler = ENAS_Scheduler( mynet, train_set=train_set, val_set=val_set,
                                reward_fn=reward_fn, batch_size=batch_size, num_gpus=1,
                                warmup_epochs=0, epochs=1, controller_lr=3e-3,
                                plot_frequency=10, update_arch_frequency=5, checkname=checkname)
```
so that's it.

with the change I can train the enas using GPU successfully

```
root@ba:/app/code# python3 enas_example.py 
running
Average latency is 7.01 ms, latency of the current architecture is 8.85 ms
Number of parameters is 18410
get_built_in_dataset mnist
get_built_in_dataset mnist
  0%|                                                                                   | 0/1 [00:00<?, ?it/s[23:10:26] 
src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:97: Running performance tests to find the best convolution algorithm, this can take a while... (set the environment variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
avg reward: 1.03: 100%|█████████████████████████████████████████████████████| 468/468 [00:40<00:00, 11.69it/s]
Val Acc: 0.9824: 100%|███████████████████████████████████████████████████████| 79/79 [00:00<00:00, 127.97it/s]
epoch 0, val_acc: 0.98, avg reward: 1.03: 100%|█████████████████████████████████| 1/1 [00:40<00:00, 40.77s/it]
```

**Comment by andyDoucette** [comment]
For anyone on the net who comes here through google because of a `Illegal instruction (core dumped)` error, here's what I found out about that:  The root cause had nothing to do with my version of mxnet.  It had to do with the hardware I was running on.  When I first submitted the issue, I was working on a g3.xlarge instance in aws.  When I went to replicate it later, I apparently started a g2.2xlarge instance instead, which apparently is incompatible at some deep level with mxnet.  I didn't dive into it too much, but upgrading to more recent hardware fixed the issue.  The clue that lead me to this was when the cpu only version of mxnet==1.6.0 also gave that "Illegal instruction" when I ran `python3 -c import mxnet`, when I was on the aws server but not on my local machine.  So, using a g3.xlarge or more recent set of hardware seems to solve that one.

**Comment by andyDoucette** [comment]
@zhreshold :  Thanks so much for your help!  I tried your fix and it does indeed run now. 

Some questions for you:

1.  Since we set the context to CPU and that's the default, can I take out all of my context setting code and it's equivalent?  For example:  `    x = mx.nd.random.uniform(shape=(1, 1, 28, 28), ctx=context)
` and `.transform(lambda x,y: to_context(context, x, y)` in the get_built_in_dataset lines?
1. What is the thing that actually makes it use the gpu?  num_gpus=1?  That's it? :)
1.  Can you explain to me the "right way" to do ENAS on GPU and why it makes sense?  Right now, I have no functional mental model of what is going on so it's very hard to reason about.  My best mental model was "put everything on the GPU, and run it there" but that didn't work.  I don't understand why it did not work.  Can you help with that?

**Comment by zhreshold** [comment]
@andyDoucette for 1&2, yes you can skip the context because the default context is mx.cpu, and what actually makes it use gpu is the `num_gpus` arg. 

For 3, your intuition is no doubt correct, "GPU is more powerful, let's put everything in GPU for training". However, nvidia doesn't allow us to `fork(happens when python multiprocessing is involved)` arrays inside GPU context(cross-process ownership of memory chunk on GPU is not allowed, aka CUDA: initialization error). 

Here's trick of ENAS, which basically assumes you always intialize parameters/data on CPU, once the `fork` is finished, use the subprocess to copy data to GPU, the ownership of the new cuda array belongs to the subprocess rather than main process, so we are good.

**Comment by andyDoucette** [comment]
Great, thank you very much for this!  Is there a place this knowledge can go in the documentation so it doesn't get lost?  And/or a check that can be done in autogluon to prevent anyone seeing a ever-not-helpful "terminate called after throwing an instance of 'dmlc::Error'" message again?  I've been a developer for years and there are friendly error message (python stack traces) and there are not so friendly error messages (stack traces pointing to un-labeled lines of 3rd party compiled libraries).  This was definitely of the "not-so-friendly" variety.

Once again, you're a life saver, thanks so much for your help! :smile:

**Comment by zhreshold** [comment]
good idea, I think a context checker can be applied to the ENAS when network is allocated on gpu but needs multiprocessing.

**Comment by uzhao** [comment]
So is multiprocessing also reason for TabularPrediction?

**Comment by Innixma** [comment]
> So is multiprocessing also reason for TabularPrediction?

I think you have separate issue. Could you create a new issue highlighting your problem with code and error logs?

**Comment by Innixma** [comment]
Closing as it appears the issue is resolved.
